File: architecture-skills/object-oriented-programming-paradigm/examples/liskov-substitution/liskov-substitution.java
// Liskov Substitution Principle in Java.
//
// Square-extends-Rectangle is the classic violation. The fix keeps
// Rectangle and Square as separate immutable types behind a Shape interface.

class LiskovSubstitution {

    // --- Violation ---

    static class Rectangle {
        protected double width, height;

        Rectangle(double w, double h) {
            width = w;
            height = h;
        }

        void setWidth(double w) { width = w; }
        void setHeight(double h) { height = h; }
        double area() { return width * height; }
    }

    static final class Square extends Rectangle {
        Square(double side) { super(side, side); }

        @Override void setWidth(double w) { width = height = w; }
        @Override void setHeight(double h) { width = height = h; }
    }

    static double stretchTo4x5(Rectangle r) {
        r.setWidth(4);
        r.setHeight(5);
        return r.area();
    }

    // --- Fix ---

    interface Shape {
        double area();
    }

    record Rect(double width, double height) implements Shape {
        public double area() { return width * height; }
    }

    record Sq(double side) implements Shape {
        public double area() { return side * side; }
    }

    static double totalArea(Shape[] shapes) {
        double sum = 0;
        for (Shape s : shapes) sum += s.area();
        return sum;
    }

    public static void main(String[] args) {
        System.out.println("-- Violation: Square substituted for Rectangle --");
        for (Rectangle rect : new Rectangle[] { new Rectangle(2, 3), new Square(2) }) {
            double got = stretchTo4x5(rect);
            String verdict = got == 20 ? "ok" : "CONTRACT BROKEN";
            System.out.printf("%9s: expected 20, got %g  (%s)%n",
                    rect.getClass().getSimpleName(), got, verdict);
        }

        System.out.println("\n-- Fix: separate types behind Shape --");
        Shape[] shapes = { new Rect(4, 5), new Sq(5) };
        for (Shape shape : shapes) {
            System.out.printf("%9s: area %g%n", shape.getClass().getSimpleName(), shape.area());
        }
        System.out.printf("    total: %g%n", totalArea(shapes));
    }
}

File: architecture-skills/object-oriented-programming-paradigm/examples/liskov-substitution/liskov-substitution.swift
// Liskov Substitution Principle in Swift.
//
// Square-extends-Rectangle is the classic violation. The fix keeps
// Rectangle and Square as separate immutable types behind a Shape protocol.

import Foundation

// --- Violation ---

class Rectangle {
    var width: Double
    var height: Double

    init(width: Double, height: Double) {
        self.width = width
        self.height = height
    }

    func setWidth(_ w: Double) { width = w }
    func setHeight(_ h: Double) { height = h }
    func area() -> Double { width * height }
}

final class Square: Rectangle {
    init(side: Double) { super.init(width: side, height: side) }

    override func setWidth(_ w: Double) { width = height = w }
    override func setHeight(_ h: Double) { width = height = h }
}

func stretchTo4x5(_ rect: Rectangle) -> Double {
    rect.setWidth(4)
    rect.setHeight(5)
    return rect.area()
}

// --- Fix ---

protocol Shape {
    func area() -> Double
}

struct Rect: Shape {
    let width: Double
    let height: Double
    func area() -> Double { width * height }
}

struct Sq: Shape {
    let side: Double
    func area() -> Double { side * side }
}

func totalArea(_ shapes: [Shape]) -> Double {
    shapes.reduce(0) { $0 + $1.area() }
}

print("-- Violation: Square substituted for Rectangle --")
for rect: Rectangle in [Rectangle(width: 2, height: 3), Square(side: 2)] {
    let got = stretchTo4x5(rect)
    let verdict = got == 20 ? "ok" : "CONTRACT BROKEN"
    print(String(format: "%9@: expected 20, got %g  (%@)", String(describing: type(of: rect)), got, verdict))
}

print("\n-- Fix: separate types behind Shape --")
let shapes: [Shape] = [Rect(width: 4, height: 5), Sq(side: 5)]
for shape in shapes {
    print(String(format: "%9@: area %g", String(describing: type(of: shape)), shape.area()))
}
print(String(format: "    total: %g", totalArea(shapes)))
